src/envoy/utils/utils.cc

```diff
@@ -34,6 +34,29 @@ const std::string kPerHostMetadataKey("istio");
 // Attribute field for per-host data override
 const std::string kMetadataDestinationUID("uid");
 
+bool getCertSAN(const Network::Connection* connection, bool peer,
+                std::string* principal) {
```

Comment:
You didn't check if principal is nullptr so probably you should declare it as std::string& principal

Reply:
That's what I was wondering as well in previous comments. It seems not just this method is passing raw pointer to the method and assume caller's responsible to ensure the memory safety. other methods around doing similar things.

Wondering what are general C++/Envoy's style guide in these cases? non const reference?

Reply:
>That's what I was wondering as well in previous comments.

:+1: 

> non const reference?

That is what I know

For this particular function as <bool, string> is the output, the absl::optional<std::string> is recommended

Reply:
I literally copied the old code, I think this is just okay because it's simply an out param, there are a lot of API having the same assumptions that the caller is going to set up all the out param before calling the function. I can change the newly added functions but will leave the old code as-is in this PR.

Reply:
If this the old convention, you can leave it here and create a issue to address it in the future :)

Reply:
@yangminzhu  Sorry I didn't look into the folded comment!